dozer/cogs/filter.py

```diff
@@ -0,0 +1,256 @@
+import discord
+import re
+from discord.ext.commands import guild_only, has_permissions
+import datetime
+
+from ._utils import *
+from .. import db
+
+
+class Filter(Cog):
+    """The filters need to be compiled each time they're run, but we don't want to compile every filter
+    Every time it's run, or all of them at once when the bot starts. So the first time that filter is run,
+    the compiled object is placed in here. This dict is actually a dict full of dicts, with each parent dict's key
+    being the guild ID for easy accessing.
+    """
+    filter_dict = {}
+
+    """Helper Functions"""
+
+    async def check_dm_filter(self, ctx, embed):
+        with db.Session() as session:
+            results = session.query(WordFilterSetting).filter_by(guild_id=ctx.guild.id, setting_type="dm") \
+                .one_or_none()
+
+            if results is None:
+                results = True
+            else:
+                results = results.value
+
+            if results is "1":
+                await ctx.author.send(embed=embed)
+                await ctx.message.add_reaction("📬")
+            else:
+                await ctx.send(embed=embed)
+
+    def load_filters(self, guild_id):
+        with db.Session() as session:
+            results = session.query(WordFilter).filter_by(guild_id=guild_id, enabled=True).all()
+            self.filter_dict[guild_id] = {}
+            for filter in results:
+                self.filter_dict[guild_id][filter.id] = re.compile(filter.pattern, re.IGNORECASE)
+
+    async def check_filters(self, message):
+        if message.author.id == self.bot.user.id:
+            return
+        with db.Session() as session:
+            roles = session.query(WordFilterRoleWhitelist).filter_by(guild_id = message.guild.id).all()
+        whitelisted_ids = set(role.role_id for role in roles)
+        if any(x.id in whitelisted_ids for x in message.author.roles):
+            return
+        filters = {}
+        try:
+            filters = self.filter_dict[message.guild.id]
+        except KeyError:
+            self.load_filters(message.guild.id)
+            filters = self.filter_dict[message.guild.id]
+        for id, filter in filters.items():
+            if filter.search(message.content) is not None:
+                await message.channel.send("{}, Banned word detected!".format(message.author.mention), delete_after=5.0)
+                time = datetime.datetime.utcnow()
+                with db.Session() as session:
+                    infraction = WordFilterInfraction(member_id=message.author.id, filter_id=id,
+                                                      timestamp=time,
+                                                      message=message.content)
+                    session.add(infraction)
+                await message.delete()
```

Comment:
If a message matches multiple bans, it will attempt to be deleted multiple times, will raise a `NotFound`.